ethereum/contracts/Messages.sol

```diff
@@ -6,17 +6,56 @@ pragma experimental ABIEncoderV2;
 
 import "./Getters.sol";
 import "./Structs.sol";
-import "./libraries/external/BytesLib.sol";
+import "./libraries/relayer/BytesParsing.sol";
 
 
 contract Messages is Getters {
-    using BytesLib for bytes;
+    using BytesParsing for bytes;
+
+    function parseAndVerifyVMOptimized(
+        bytes calldata encodedVM,
+        bytes calldata guardianSet,
+        uint32 guardianSetIndex
+    ) public view returns (Structs.VM memory vm, bool valid, string memory reason) {
+        // Verify that the specified guardian set is a valid.
+        require(
+            getGuardianSetHash(guardianSetIndex) == keccak256(guardianSet),
+            "invalid guardian set"
+        );
+
+        vm = parseVM(encodedVM);
+
+        // Verify that the VM is signed with the same guardian set that was specified.
+        require(vm.guardianSetIndex == guardianSetIndex, "mismatched guardian set index");
+
+        (valid, reason) = verifyVMInternal(vm, parseGuardianSet(guardianSet), false);
+    }
+
+    function parseGuardianSet(bytes calldata guardianSetData) public pure returns (Structs.GuardianSet memory guardianSet) {
+        // Fetch the guardian set length.
+        uint256 endGuardianKeyIndex = guardianSetData.length - 4;
+        uint256 guardianCount = endGuardianKeyIndex / 20;
```

Comment:
I think it would be nice to specify that this `20` is the size of an address, perhaps with an `ADDRESS_SIZE` constant or something equivalent.

Reply:
Agreed.